mnefun/_mnefun.py

```diff
@@ -2330,3 +2330,117 @@ def plot_reconstruction(evoked, origin=(0., 0., 0.04)):
     axs[2, 0].set_ylabel('Reconstruction')
     fig.tight_layout()
     return fig
+
+
+def chpi_snr_raw(raw, winlen, nharm, stop=None):
+    """ Compute and plot cHPI SNR from raw data.
+
+    Parameters
+    ----------
+
+    winlen : int
+        Length of window to use. Longer windows will naturally include more
+        low frequency power, resulting in lower SNR.
+
+    nharm : int
+        Number of line frequency harmonics to include in the model.
+
+    stop : int
+        If specified, stop before the end of file (seconds)
+    """
+
+    # plotting parameters
+    legend_fontsize = 12
+    legend_hspace = 30  # % of horizontal space to reserve for legend
+
+    # get some info from fiff
+    sfreq = raw.info['sfreq']
+    linefreq = raw.info['line_freq']
+    linefreqs = (np.arange(nharm + 1) + 1) * linefreq
+    buflen = int(winlen * sfreq)
+    if buflen <= 0:
+        raise ValueError('Window length should be >0')
+    cfreqs = []
+    if (len(raw.info['hpi_meas']) > 0 and
+            'coil_freq' in raw.info['hpi_meas'][0]['hpi_coils'][0]):
+        for coil in raw.info['hpi_meas'][0]['hpi_coils']:
+            cfreqs.append(coil['coil_freq'][0])
+    else:
+        raise ValueError('Cannot determine cHPI frequencies'
+                         'from raw data info')
+    print('\nNominal cHPI frequencies: ', cfreqs, ' Hz')
+    print('Sampling frequency:', sfreq, 'Hz')
+    print('Using line freqs:', linefreqs, ' Hz')
+    print('Using buffers of', buflen, 'samples =', buflen/sfreq, 'seconds\n')
+
+    raw.info['bads'] = []  # make sure to load all channels
+    pick_meg = pick_types(raw.info, meg=True)
+    pick_mag = pick_types(raw.info, meg='mag')
+    pick_grad = pick_types(raw.info, meg='grad')
+    nchan = len(pick_meg)
+
+    # create general linear model for the data
+    t = np.linspace(0, buflen/sfreq, endpoint=False, num=buflen)
+    model = np.c_[t, np.ones(t.shape)]  # model slope and DC
+    # add sine and cosine term for each freq
+    for f in list(linefreqs) + cfreqs:
+        model = np.c_[model, np.cos(2*np.pi*f*t), np.sin(2*np.pi*f*t)]
+    inv_model = np.linalg.pinv(model)
+
+    # loop through MEG data, fit linear model and compute SNR at each window
+    if stop:
+        stop = int(stop*sfreq)
+    else:
+        stop = raw.n_times
+    # drop last buffer to avoid overrun
+    bufs = range(0, int(stop), buflen)[:-1]
```

Comment:
... and `range` on Py3k is not a list but a generator, so it's better to use `np.arange` instead